services/thingsearch/persistence/src/main/java/org/eclipse/ditto/services/thingsearch/persistence/write/streaming/EnforcementFlow.java

```diff
@@ -298,4 +315,76 @@ private static boolean shouldReloadCache(@Nullable final Entry<?> entry, final M
         return lazySource.viaMat(Flow.create(), Keep.none());
     }
 
+    /**
+     * Get the imports of a thing's policy or an empty source if it does not exist.
+     *
+     * @param metadata metadata of the thing.
+     * @param thing    the thing (possibly containing ACL)
+     * @return source of an enforcer or an empty source.
+     */
+    private Source<List<Entry<Policy>>, NotUsed> getPolicyImports(final Metadata metadata, final JsonObject thing) {
+        final Optional<JsonObject> acl = thing.getValue(Thing.JsonFields.ACL);
+        if (acl.isPresent()) {
+            return POLICY_IMPORTS_NONEXISTENT;
+        } else {
+            try {
+                return thing.getValue(Thing.JsonFields.POLICY_ID).map(PolicyId::of)
+                        .map(policyId -> readCachedImportedPolicies(metadata, getPolicyEntityId(policyId), 0))
+                        .orElse(POLICY_IMPORTS_NONEXISTENT);
+            } catch (PolicyIdInvalidException e) {
+                return POLICY_IMPORTS_NONEXISTENT;
+            }
+        }
+    }
+
+    private Source<Entry<Policy>, NotUsed> readCachedPolicy(final Metadata metadata,
+            final EntityIdWithResourceType policyId, final int iteration) {
+
+        final Source<Entry<Policy>, ?> lazySource = Source.lazySource(() -> {
+            final CompletionStage<Source<Entry<Policy>, NotUsed>> policyFuture = policyCache.get(policyId)
+                    .thenApply(optionalPolicyEntry -> {
+                        if (shouldReloadCache(optionalPolicyEntry.orElse(null), metadata, iteration)) {
+                            // invalid entry; invalidate and retry after delay
+                            policyCache.invalidate(policyId);
+                            return readCachedPolicy(metadata, policyId, iteration + 1)
+                                    .initialDelay(cacheRetryDelay);
+                        } else {
+                            return optionalPolicyEntry.map(Source::single)
+                                    .orElse(POLICY_NONEXISTENT);
+                        }
+                    })
+                    .exceptionally(error -> {
+                        log.error("Failed to read policyCache", error);
+                        return POLICY_NONEXISTENT;
+                    });
+
+            return Source.completionStageSource(policyFuture);
+        });
+
+        return lazySource.viaMat(Flow.create(), Keep.none());
+    }
+
+    private Source<List<Entry<Policy>>, NotUsed> readCachedImportedPolicies(final Metadata metadata,
```

Comment:
@thjaeckle I think this is somewhat like you meant on gitter?

Reply:
@vhdirk hm, almost :)
I don't get why you load the complete policy of the "imported policies" - for the `List<PolicyTag> policyImports`  passed into the `EnforcedThingMapper.toWriteModel(..)` method you don't need to load the `Policy`, the id of the imported policies + the revision number is sufficient.
I would simplify the return types of the private "read..." methods. Also, I don't think that you have to use `Entry` there as wrapping cache entry type - just use the `PolicyTag` or the list of dertermined `PolicyTag`s.

I would expect a method signature:
```
private Source<List<PolicyTag>, NotUsed> getPolicyImports(final PolicyId policyId);
```
Which can simply be called in addition to `getEnforcer` to get input for the to be called `EnforcedThingMapper.toWriteModel()`.

Reply:
Ok, I'll refactor it. You weren't kidding when you said that this requires messing about with some deep internals :p

Reply:
I suppose it also doesn't quite make sense to reload the cached entry? https://github.com/vhdirk/ditto/blob/7d57d4f590300b14d242cdb299d31beaf04c90db/services/thingsearch/persistence/src/main/java/org/eclipse/ditto/services/thingsearch/persistence/write/streaming/EnforcementFlow.java#L362

Reply:
While refactoring, I think I do need to load at least the thing policy; it may be that a cached version of a policy that is imported by the 'current' policy already exists. We won't know that unless we fetch the latest revision of the 'current' policy, or do we?

Reply:
Good point 👍 
The `Metadata` object contains the revision number of the used policy of a Thing update processed in `ThingUpdater`.
However at this point it can't be known whether any of the referenced imported policies have also changed. Even if all cache invalidator callbacks would work as expected, due to the sharding when horizontally scaling the "things-search" service, this information is not consistently known everywhere.

So I think you're right: when a policy uses imported policies, the imported policies (to be precise: their revision number) must be loaded OR the cache entries in the `policyCache` for those must be invalidated which will cause an automatic new loading of them.

This would make it a very expensive (bandwith and CPU consumption wise for deserialization) operation via the "normal" `policyCache` - as this cache always asks for the full Policy JSON.

TL;DR: So I fear we need another cache which holds `PolicyTag`s.

Reply:
And for such a new cache to work, I guess we're also going to need a new `sudo` command, `SudoRetrievePolicyRevision` or so? Or, perhaps `SudoRetrievePolicyRevisions`, since we know it's going to be a list.

Reply:
Yes. You mean like this already existing one `org.eclipse.ditto.services.models.policies.commands.sudo.SudoRetrievePolicyRevision`? ;)

Reply:
And to know I did actually look for something like that. Time for more coffee :)
Aside from that; this MR is getting pretty big. Do you think we could/need split it up somehow?

Reply:
I don't think that splitting up is useful - we wouldn't want to merge a half-working feature to our master branch.
Documentation and integration tests could be provided in another PR afterwards.

And with Ditto 2.0 to be released next week, there will be some (huge, I fear) merge efforts as well on the overall PR - I can support on that.